examples/grad3d/grad3d.f90

```diff
@@ -0,0 +1,625 @@
+   ! This operation is also done as part of decomp_2d_init
+   call MPI_COMM_SIZE(MPI_COMM_WORLD, nranks_tot, ierror)
+   resize_domain = int(nranks_tot / 4) + 1
+   nx = nx_base * resize_domain
+   ny = ny_base * resize_domain
+   nz = nz_base * resize_domain
+   ! Now we can check if user put some inputs
+   ! Handle input file like a boss -- GD
+   nargin = command_argument_count()
+   if ((nargin == 0) .or. (nargin == 2) .or. (nargin == 5)) then
+      do arg = 1, nargin
+         call get_command_argument(arg, InputFN, FNLength, status)
+         read (InputFN, *, iostat=status) DecInd
+         if (arg == 1) then
+            p_row = DecInd
+         elseif (arg == 2) then
+            p_col = DecInd
+         elseif (arg == 3) then
+            nx = DecInd
+         elseif (arg == 4) then
+            ny = DecInd
+         elseif (arg == 5) then
+            nz = DecInd
+         end if
+      end do
+   else
+      ! nrank not yet computed we need to avoid write
+      ! for every rank
+      call MPI_COMM_RANK(MPI_COMM_WORLD, nrank, ierror)
+      if (nrank == 0) then
+         print *, "This Test takes no inputs or 2 inputs as"
+         print *, "  1) p_row (default=0)"
+         print *, "  2) p_col (default=0)"
+         print *, "or 5 inputs as"
+         print *, "  1) p_row (default=0)"
+         print *, "  2) p_col (default=0)"
+         print *, "  3) nx "
+         print *, "  4) ny "
+         print *, "  5) nz "
+         print *, "Number of inputs is not correct and the defult settings"
+         print *, "will be used"
+      end if
+   end if
+   call decomp_2d_init(nx, ny, nz, p_row, p_col)
+
+   dx = lx/real(nx,mytype)
+   dy = ly/real(ny,mytype)
+   dz = lz/real(nz,mytype)
+   
+   if (nrank == 0) then
+      write (*, *) '-----------------------------------------------'
+      write (*, *) "Mesh Resolution ", nx , ny, nz 
+      write (*, *) '-----------------------------------------------'
+   end if
+
+   call allocate_var()
+   !$acc data create(phi2,phi3,wk2,wk3) copy(phi1,dphiX, dphiY, dphiZ)
+   call init_phi()
+   call compute_grad()
+   call test_derX(dphiX)
+   call test_derY(dphiY)
+   call test_derZ(dphiZ)
+   call write_data()
+   !$acc end data
+
+   if (nrank == 0) then
+      write (*, *) '-----------------------------------------------'
+      write (*, *) "End GRAD calculation check all pass: ", all_pass
+      write (*, *) '==============================================='
+   end if
+
+   call finalize()
+
+   call decomp_2d_finalize
+
+   if (.not. all_pass) call decomp_2d_abort(1, "Error in grad3d")
+
+   call MPI_FINALIZE(ierror)
+
+contains
+
+   !=====================================================================
+   ! Initialize
+   !=====================================================================
+   subroutine allocate_var()
+
+      use decomp_2d
+
+      implicit none
+
+      logical, parameter :: glob = .false.
+
+      ! initialise u,v,w with random numbers in X-pencil
```

Comment:
! initialise phi with 3D harmonic function

Reply:
Change with `! Allocate main variables in X-pencil`